File: Specializations/Data Structures and Algorithms/Data Structures/Week 5/Assignment/src/main/scala/TreeOrders.scala
import java.io.{BufferedReader, IOException, InputStream, InputStreamReader}
import java.util.StringTokenizer

object TreeOrders {

  abstract class BinaryTree extends {
    override def toString: String = {
      def toStringIter(tree: BinaryTree): String =
        tree match {
          case Empty() => "- "
          case Node(k, l, r) => "Node(" + k + ", " + toStringIter(l) + ", " + toStringIter(r) + ")"
          case Leaf(k) => "Leaf(" + k + ")"
        }
      toStringIter(this)
    }
  }

  def inOrderTraversal(tree: BinaryTree, sb: StringBuilder): StringBuilder = tree match {
    case Empty() => sb
    case Node(k, l, r) => {
      inOrderTraversal(l, sb)
      sb.append(k + " ")
      inOrderTraversal(r, sb)
    }
    case Leaf(k) => sb.append(k + " ")
  }

  def preOrderTraversal(tree: BinaryTree, sb: StringBuilder): StringBuilder = tree match {
    case Empty() => sb
    case Node(k, l, r) => {
      sb.append(k + " ")
      preOrderTraversal(l, sb)
      preOrderTraversal(r, sb)
    }
    case Leaf(k) => sb.append(k + " ")
  }

  def postOrderTraversal(tree: BinaryTree, sb: StringBuilder): StringBuilder = tree match {
    case Empty() => sb
    case Node(k, l, r) => {
      postOrderTraversal(l, sb)
      postOrderTraversal(r, sb)
      sb.append(k + " ")
    }
    case Leaf(k) => sb.append(k + " ")
  }

  case class Empty() extends BinaryTree
  case class Leaf(key: Int) extends BinaryTree
  case class Node(key: Int, left: BinaryTree, right: BinaryTree) extends BinaryTree

  def buildBinaryTree(keys: Array[Int], lefts: Array[Int], rights: Array[Int]): BinaryTree = {
    def buildBinaryTreeFromIndexes(key: Int, left: Int, right: Int): BinaryTree =
      (key, left, right) match {
        case (k, l, r) if l > -1 && r > -1 =>
          Node(k, buildBinaryTreeFromIndexes(keys(l), lefts(l), rights(l)),
            buildBinaryTreeFromIndexes(keys(r), lefts(r), rights(r)))
        case (k, -1, r) if r > -1 =>
          Node(k, Empty(), buildBinaryTreeFromIndexes(keys(r), lefts(r), rights(r)))
        case (k, l, -1) if l > -1 =>
          Node(k, buildBinaryTreeFromIndexes(keys(l), lefts(l), rights(l)), Empty())
        case (k, -1, -1) => Leaf(k)
      }

    def findStartIndex(index: Int, left: Int, right: Int): Int = {
      if (left == -1 && right == -1) findStartIndex(index + 1, lefts(index + 1), rights(index + 1))
      else index
    }

    val startIndex = findStartIndex(0, lefts(0), rights(0))
    buildBinaryTreeFromIndexes(keys(startIndex), lefts(startIndex), rights(startIndex))
  }

  def buildTree(n: Int, read: Unit => Int): BinaryTree = {
    val keys = Array.ofDim[Int](n)
    val left = Array.ofDim[Int](n)
    val right = Array.ofDim[Int](n)
    for (i <- 0 until n) {
      keys(i) = read()
      left(i) = read()
      right(i) = read()
    }

    buildBinaryTree(keys, left, right)
  }

  def main(args: Array[String]): Unit = {
    new Thread(null, new Runnable() {
      def run() {
        try
          runTree()

        catch {
          case e: IOException => {
          }
        }
      }
    }, "1", 1 << 26).start()
  }

  @throws[IOException]
  def runTree() {
    val scanner: FastScanner = new FastScanner(System.in)
    val n: Int = scanner.nextInt

    val t = buildTree(n, _ => scanner.nextInt)
    println(inOrderTraversal(t, new StringBuilder()).toString().trim)
    println(preOrderTraversal(t, new StringBuilder()).toString().trim)
    println(postOrderTraversal(t, new StringBuilder()).toString().trim)
  }

  class FastScanner(val stream: InputStream) {

    var br: BufferedReader = null
    var st: StringTokenizer = null

    try
      br = new BufferedReader(new InputStreamReader(stream))

    catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }

    def next: String = {
      while (st == null || !st.hasMoreTokens)
        try
          st = new StringTokenizer(br.readLine)

        catch {
          case e: IOException => {
            e.printStackTrace()
          }
        }
      st.nextToken
    }

    def nextInt: Int = next.toInt
    def nextLong: Long = next.toLong
  }
}

File: Specializations/Data Structures and Algorithms/Data Structures/Week 5/Assignment/src/main/java/tree_orders.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.StringTokenizer;

/**
 * Binary tree traversals.
 * <p>
 * Remember when to push node into the stack
 * and when to pop node out of the stack.
 */
public class tree_orders {
    class FastScanner {
        StringTokenizer tok = new StringTokenizer("");
        BufferedReader in;

        FastScanner() {
            in = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() throws IOException {
            while (!tok.hasMoreElements())
                tok = new StringTokenizer(in.readLine());
            return tok.nextToken();
        }

        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }

    public class TreeOrders {
        int n;
        int[] key, left, right;

        void read() throws IOException {
            FastScanner in = new FastScanner();
            n = in.nextInt();
            key = new int[n];
            left = new int[n];
            right = new int[n];
            for (int i = 0; i < n; i++) {
                key[i] = in.nextInt();
                left[i] = in.nextInt();
                right[i] = in.nextInt();
            }
        }

        /**
         * Key Idea:
         * Start at root vertex 0.
         * Push all nodes `along the path` from root to left-most into stack.
         * While stack is not empty, print top.
         * Push all nodes along the path from its right-subtree to left-most.
         *
         * @return
         */
        List<Integer> inOrder() {
            ArrayList<Integer> result = new ArrayList<Integer>();
            Stack<Integer> stack = new Stack<>();
            // Refactor out the common code
            pushAll(stack, 0);
            while (! stack.isEmpty()) {
                int top = stack.pop();
                result.add(key[top]);
                top = right[top];  // Move pointer to right subtree
                pushAll(stack, top);
            }
            return result;
        }
        /** Push all nodes on the path root->left-most into stack */
        private void pushAll(Stack<Integer> stack, int root) {
            while (root != -1) {
                stack.push(root);
                root = left[root];
            }
        }

        /**
         * Key Idea:
         * Push root into stack.
         * While stack is not empty, pop & print top.
         * Push its non-null right, left child into stack.
         *
         * @return
         */
        List<Integer> preOrder() {
            List<Integer> result = new ArrayList<Integer>();
            Stack<Integer> stack = new Stack<>();
            stack.push(0);
            while (! stack.isEmpty()) {
                int top = stack.pop();
                result.add(key[top]);
                if (right[top] != -1) stack.push(right[top]);
                if (left[top]  != -1) stack.push(left[top]);
            }
            return result;
        }

        /**
         * Key Idea: add sentinel? Still PreOrder</br>
         * Given an output of postOrder, what is from right to left?
         * <pre> Still PreOrder, but right side first.. </pre>
         * Push root and while stack is not empty, pop top.
         * Insert top at the head (reverse-order).
         * Push its non-null left, right child into stack.
         *
         * @return
         */
        List<Integer> postOrder() {
            ArrayList<Integer> result = new ArrayList<Integer>();
            Stack<Integer> stack = new Stack<>();
            stack.push(0);
            while (! stack.isEmpty()) {
                int top = stack.pop();
                result.add(0, key[top]);
                // BZ: add top's children, instead of root
                if (left[top]  != -1) stack.push(left[top]);
                if (right[top] != -1) stack.push(right[top]);
            }
            return result;
        }
    }

    static public void main(String[] args) throws IOException {
        new Thread(null, new Runnable() {
            public void run() {
                try {
                    new tree_orders().run();
                } catch (IOException e) {
                }
            }
        }, "1", 1 << 26).start();
    }

    public void print(List<Integer> x) {
        for (Integer a : x) {
            System.out.print(a + " ");
        }
        System.out.println();
    }

    public void run() throws IOException {
        TreeOrders tree = new TreeOrders();
        tree.read();
        print(tree.inOrder());
        print(tree.preOrder());
        print(tree.postOrder());
    }
}